### PR TITLE
Add some specifics on gems (and maybe pips)

### PR DESCRIPTION
**Pull request by jaymzh**
In https://github.com/Homebrew/homebrew-core/pull/157910 we discussed
some improvements to docs on setting up gems. THis is an attempt at some
docs for that.

If someone can help with pip docs for this I'll add it too.

Signed-off-by: Phil Dibowitz <phil@ipom.com>

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

File: docs/Formula-Cookbook.md [hunk 1]
```diff
@@ -251,7 +251,36 @@ uses_from_macos "curl", since: :monterey
...
+And install any bins, and munge their shebang lines, with:
+
+```ruby
+bin.install libexec/"bin/<bin>"
+bin.env_script_all_files(libexec/"bin", GEM_HOME: ENV["GEM_HOME"])
+```
+
+### Python dependencies
+
+something about requirements.txt ?
```
Comment (MikeMcQuaid) [inline docs/Formula-Cookbook.md:279]: @Homebrew/maintainers can I get some help filling this in?

File: docs/Formula-Cookbook.md [hunk 2]
```diff
@@ -251,7 +251,36 @@ uses_from_macos "curl", since: :monterey
 
 Homebrew doesn’t package already-packaged language-specific libraries. These should be installed directly from `gem`/`cpan`/`pip` etc.
 
-If you're installing an application then use [`resource`](https://rubydoc.brew.sh/Formula#resource-class_method)s for all language-specific dependencies:
+### Ruby Gem Dependencies
+
+The preferred mechanism for installing gem dependencies is to use `bundler` with the upstream's `Gemfile.lock`. This requires the upstream checks in their `Gemfile.lock`, so if they don't, it's a good idea to file a bug and ask them to do so. Assuming they have one, this is as simple as:
```
Comment (MikeMcQuaid) [inline docs/Formula-Cookbook.md:256]:
```suggestion
The preferred mechanism for installing gem dependencies is to use `bundler` with the upstream's `Gemfile.lock`. This requires the upstream checks in their `Gemfile.lock`, so if they don't, it's a good idea to file an issue and ask them to do so. Assuming they have one, this is as simple as:
```

File: docs/Formula-Cookbook.md [hunk 3]
```diff
@@ -251,7 +251,36 @@ uses_from_macos "curl", since: :monterey
...
+
+```ruby
+ENV["GEM_HOME"] = libexec
+system "bundle", "install"
```
Comment (MikeMcQuaid) [inline docs/Formula-Cookbook.md:260]:
```suggestion
system "bundle", "install", "--without", "development"
```

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks for this @jaymzh, much appreciated and looking great so far.

File: docs/Formula-Cookbook.md (hunk 1 above)
Comment (cho-m) [reply docs/Formula-Cookbook.md:279]: Just noting that Python is a bit trickier as there isn't a standardized lock.
- requirements.txt isn't used everywhere and I've seen many projects that just manually modify it so not really at the same level as tooling generated equivalents (e.g. ruby Gemfile.lock, go go.mod+go.sum, rust Cargo.lock).
- Poetry-based projects have poetry.lock but Poetry isn't adopted everywhere.
- Our current approach of resources is probably the most applicable and able to result in reproducible builds.

Given our current infra/DSL, I think Python dependencies would be covered under following other cases.

File: docs/Formula-Cookbook.md (hunk 3 above)
Comment (cho-m) [reply docs/Formula-Cookbook.md:260]: On side note, perhaps these should be included as commented out lines in `brew create --ruby`. I feel that is often the easiest way to showcase our preferences (though I don't know statistics of how many people actually use that command).

Comment (jaymzh) [reply docs/Formula-Cookbook.md:260]: I didn't even know there WAS a --python and a --ruby. I'll add that to the docs in this PR.

File: docs/Formula-Cookbook.md (hunk 1 above)
Comment (jaymzh) [reply docs/Formula-Cookbook.md:279]: Updated to incorporate this in and restructure slightly.

**Comment by jaymzh** [comment]
Oh I'm always stoke to see folks using MDL linters (I'm the primary maintainer these days)!

**Comment by MikeMcQuaid** [comment]
> Oh I'm always stoke to see folks using MDL linters (I'm the primary maintainer these days)!

Oh, great: thanks for your work on such a brilliant tool!

**Review by MikeMcQuaid — APPROVED** [review]
Looks good, thanks again @jaymzh, much appreciated!

**Pull request merged**